### PR TITLE
[BraveNews]  Fix the dialog not closing when clicking Brave news button while it's showing.

### PR DESCRIPTION
**Pull request by jagadeshjai**
- Refactor Brave news's button and dialog.
- Fix the dialog not closing when clicking brave news button while it's showing.

**Note:** 
Basically, this issue is fixed when we use the `OnWidgetDestroyed` method of `BraveNewsBubbleView` to clear the current bubble (slower, so it will not notify(i.e `OnExecuting`) since the bubble is in the showing state) instead of depending on `SetCloseCallback` (faster, but the bubble is in not showing state which triggers notify event, so the bubble will reshow).

Refer: https://source.chromium.org/chromium/chromium/src/+/main:chrome/browser/ui/views/location_bar/icon_label_bubble_view.h;l=291;drc=f2263729318bf6737e153bc27e70ab3379a789b4

Resolves https://github.com/brave/brave-browser/issues/37216

Outcome:
[brave_news_button_dialog_closing.webm](https://github.com/brave/brave-core/assets/45998266/24c1bdad-e584-4e5a-ae13-82281b3ba479)

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

**Comment by jagadeshjai** [comment]
@fallaciousreasoning @szilardszaloki 

Kindly review this PR.

File: browser/ui/BUILD.gn [hunk 1]
```diff
@@ -1047,6 +1047,10 @@ source_set("ui") {
       "views/brave_help_bubble/brave_help_bubble_delegate_view.h",
       "views/brave_help_bubble/brave_help_bubble_host_view.cc",
       "views/brave_help_bubble/brave_help_bubble_host_view.h",
+      "views/brave_news/brave_news_bubble_controller.cc",
+      "views/brave_news/brave_news_bubble_controller.h",
+      "views/brave_news/brave_news_location_view.cc",
```
[the remark this answers is not in the document]
Comment (fallaciousreasoning) [reply browser/ui/BUILD.gn:1052]: Yeah, I agree that'd make more sense but then again, the other icons Brave has up there are called `location_view` too. Maybe we just leave this one for now?

[the remark this answers is not in the document]
Comment (fallaciousreasoning) [reply browser/ui/BUILD.gn:1052]: Oh just saw it moved - yeah, let's rename it :smile:

**Review by fallaciousreasoning — COMMENTED** [review]
Looks great! I'm going to kick off the tests

**Comment by fallaciousreasoning** [comment]
Kicking off the tests again - I'll merge if the build passes! Thanks heaps @jagadeshjai

**Comment by jagadeshjai** [comment]
@fallaciousreasoning From the Security Error, I guess you need to create a duplicate CL to trigger test from your account.

**Comment by fallaciousreasoning** [comment]
Oh man, I thought we'd fixed that. So annoying :laughing: - kicked off #24274

**Comment by fallaciousreasoning** [comment]
@jagadeshjai presubmit failed on CI - do you mind running `npm run presubmit -- --fix` and fixing any errors?

```
01:28:44  Running Python 3 presubmit upload checks ...
01:28:54    checking for commit objects in tree took a long time: 4.2s
01:28:54  ** Presubmit Warnings: 2 **
01:28:54  browser/ui/views/brave_news/brave_news_bubble_controller.h: PlaylistActionIconView forward declaration is no longer needed
01:28:54  
01:28:54  browser/ui/views/brave_news/brave_news_location_view.h: BraveNewsBubbleView forward declaration is no longer needed
01:28:54  
01:28:54  ** Presubmit ERRORS: 1 **
01:28:54  Expected one of license headers in new files:
01:28:54  // Copyright (c) 2024 The Brave Authors. All rights reserved.
01:28:54  // This Source Code Form is subject to the terms of the Mozilla Public
01:28:54  // License, v. 2.0. If a copy of the MPL was not distributed with this file,
01:28:54  // You can obtain one at https://mozilla.org/MPL/2.0/.
01:28:54  
01:28:54  /* Copyright (c) 2024 The Brave Authors. All rights reserved.
01:28:54   * This Source Code Form is subject to the terms of the Mozilla Public
01:28:54   * License, v. 2.0. If a copy of the MPL was not distributed with this file,
01:28:54   * You can obtain one at https://mozilla.org/MPL/2.0/. */
01:28:54  
01:28:54  # Copyright (c) 2024 The Brave Authors. All rights reserved.
01:28:54  # This Source Code Form is subject to the terms of the Mozilla Public
01:28:54  # License, v. 2.0. If a copy of the MPL was not distributed with this file,
01:28:54  # You can obtain one at https://mozilla.org/MPL/2.0/.
01:28:54  
01:28:54    browser/ui/views/brave_news/brave_news_location_view.cc
01:28:54    browser/ui/views/brave_news/brave_news_location_view.h
01:28:54  
01:28:54  Presubmit checks took 8.3s to calculate.
```

**Comment by jagadeshjai** [comment]
@fallaciousreasoning Checked in the changes. PTAL.

**Comment by fallaciousreasoning** [comment]
Just rerunning the MacOS tests - it failed but looked unrelated.

**Review by fallaciousreasoning — APPROVED** [review]
Looks great - CI is passing so I'm going to merge

**Comment by fallaciousreasoning** [comment]
Sorry, looks like the signing still hasn't worked. I found https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits handy when getting it set up (it's definitely a bit faffy, sorry)

**Comment by jagadeshjai** [comment]
> Sorry, looks like the signing still hasn't worked. I found https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits handy when getting it set up (it's definitely a bit faffy, sorry)

Yes, I'm checking on it, However this doc seems only guiding for GPG key, but I'm using SSH key. So couldn't get how to sign already committed commit :smiling_face_with_tear:

**Comment by fallaciousreasoning** [comment]
Yeah, I use an ssh key too - I think you have to add it again as a signing key in the Github UI (if you haven't already). Setting `gpgsign` seems to work :shrug: . To resign a bunch of commits I normally do:

```
git rebase HEAD~2 --interactive
# edit all commits, and commit with `git commit -a -S --no-edit`
```

and hopefully that will sign them :crossed_fingers: 

Alternatively, you could do

```
git reset HEAD~2 --soft
git commit -S -am "[BraveNews] Fix the dialog not closing when clicking Brave news button while it's showing"
```

if you don't mind squashing the commits - good luck!

**Comment by jagadeshjai** [comment]
> Yeah, I use an ssh key too - I think you have to add it again as a signing key in the Github UI (if you haven't already). Setting `gpgsign` seems to work 🤷 . To resign a bunch of commits I normally do:
> 
> ```
> git rebase HEAD~2 --interactive
> # edit all commits, and commit with `git commit -a -S --no-edit`
> ```
> 
> and hopefully that will sign them 🤞


Thanks for you help @fallaciousreasoning , No luck on even after adding SSH key as a signing key. 
But setting up GPG Key worked :+1:

**Comment by fallaciousreasoning** [comment]
Okay kicked of CI again

**Comment by fallaciousreasoning** [comment]
Sorry @jagadeshjai could you rebase on latest master - looks like Android/Windows are failing (noticed they're failing on some of my personal branches too). I think the latest Chromium bump has just come through

**Comment by jagadeshjai** [comment]
> Sorry @jagadeshjai could you rebase on latest master - looks like Android/Windows are failing (noticed they're failing on some of my personal branches too). I think the latest Chromium bump has just come through

Done. There are so many unrelated failures in this PR 😂. Sorry for making you do the same work again and again.

**Comment by fallaciousreasoning** [comment]
Same to you :laughing: I think we just go unlucky with a couple of big Chromium bumps going through. This time! :crossed_fingers:

**Comment by fallaciousreasoning** [comment]
Finally got CI to pass @jagadeshjai - thanks heaps for your work on this!

**Comment by jagadeshjai** [comment]
> Finally got CI to pass @jagadeshjai - thanks heaps for your work on this!

Thanks @fallaciousreasoning :heart:

**Pull request merged**